src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -507,14 +506,50 @@ void Lowering::LowerCast(GenTree* tree)
 
     assert(!varTypeIsSmall(srcType));
 
-    if (tmpType != TYP_UNDEF)
+    // Transform '(ulong)smallInt' or '(uint)smallInt' to 'smallInt & 0xFF..'
+    if (comp->opts.OptimizationEnabled() && tree->gtGetOp1()->OperIs(GT_CAST) && !tree->gtOverflow() &&
+        tree->IsUnsigned() && !tree->isContained() && !varTypeIsSmallInt(dstType) &&
+        varTypeIsIntOrI(varTypeToSigned(dstType)))
     {
-        GenTree* tmp = comp->gtNewCastNode(tmpType, op1, tree->IsUnsigned(), tmpType);
-        tmp->gtFlags |= (tree->gtFlags & (GTF_OVERFLOW | GTF_EXCEPT));
+        GenTreeCast* cast2    = tree->gtGetOp1()->AsCast();
+        var_types    dst2Type = cast2->CastToType();
+        GenTree*     op       = cast2->gtGetOp1();
 
-        tree->gtFlags &= ~GTF_UNSIGNED;
-        tree->AsOp()->gtOp1 = tmp;
-        BlockRange().InsertAfter(op1, tmp);
+        if (varTypeIsSmallInt(dst2Type) && !cast2->gtOverflow() && !cast2->isContained() && varTypeIsIntegral(op))
+        {
+            JITDUMP("Transform zero-extended from small int via double CAST to AND 0xFF(0xFFFF). Before:\n\n")
+            DISPTREE(tree)
+
+            assert(genTypeSize(tree->CastToType()) > genTypeSize(dst2Type));
+            tree->ChangeOper(GT_AND, GenTree::PRESERVE_VN);
```

Comment:
```suggestion
            tree->ChangeOper(GT_AND);
```
In general we do not try to preserve VNs after rationalization.

Reply:
yeah I know, just a muscle memory since here it indeed is not changed after transformation